### PR TITLE
Include base image in dependency hash

### PR DESCRIPTION
**Pull request by skovhus**
https://github.com/tmrowco/tmrow/issues/3723

There is currently a flaw in the [dependency hash approach](https://github.com/tmrowco/brick/pull/46) so changes in base images (e.g. prepare) will not be reflected in down stream images (e.g. build or test images). 

The simplest solution was to utilize the Docker image id for the image that we depend on by reading the docker file content. 

### Before vs after
The dependencies for an image was: BUILD.yaml file + hash of all inputs

The dependencies for an image now is: BUILD.yaml file + hash of all inputs AND the id of the Docker input image (which will be updated if there is a new image).

### CODE REVIEW

**Review by madsnedergaard — APPROVED** [review]
Nice, did you test it with the branch by @fbarl or something similar?
Looks sane to me, haven't tested it, but seeing as this is slightly urgent/blocking I'll approve it :)

**Comment by skovhus** [comment]
> Nice, did you test it with the branch by @fbarl or something similar?

Yeah, it seems to work. But will cause some looong build time the first time as we are building all the images again. :)

File: brick/dockerlib.py
Comment (fbarl) [inline brick/dockerlib.py:192]: I think it's worth noting that this line basically prevents using [Docker Multi-stage builds](https://docs.docker.com/develop/develop-images/multistage-build/).

I'm not sure if that's the use case that we want to support, but if the answer is no, I think we should still reflect that better in the error statement, e.g. have something like this:

``` python
if len(from_image_name) < 1:
        raise Exception(f"Did not found a FROM statement in {dockerfile_contents}")
if len(from_image_name) > 1:
        raise Exception(f"Docker Multi-stage builds are not supported by brick yet: {dockerfile_contents}")
```

If we do want to support it here, we should probably concatenate all `from_image_name` strings and use that instead. And only check for `if len(from_image_name) == 0` before throwing this error of course.

WDYT?

**Review by fbarl — COMMENTED** [review]
LGTM except that one comment I left - I'd very much like it to be addressed before merging but won't block the PR in case we want to merge it urgently :wink:

File: brick/dockerlib.py
Comment (skovhus) [reply brick/dockerlib.py:192]: Good that you are bringing this up!

Note that Docker is an internal tool used by Brick, so the docker files are actually generated by Brick and the process is opaque to the user.

And as we are not using multi-stage builds, the error handling here is just to be defensive and get an error if we change something inside Brick.

So, I don't think we need to change anything. :)

Comment (fbarl) [reply brick/dockerlib.py:192]: Alright, thanks for the context, feel free to merge the PR then :slightly_smiling_face:

**Pull request merged**
